### DISCUSSION TITLE
No module named 'splink.duckdb'

### DISCUSSION THREAD

**Discussion by VaishnaviDhulipalla2902** [moj-analytical-services/splink#1127] [General]
I am trying to use Splink for record linkage and the import itself throws an error `No module named 'splink.duckdb'`. I have installed splink using anaconda - `conda install -c conda-forge splink` . How do I resolve this?

**Comment by RobinL** [comment]
What do you get if you run
```
import splink
splink.__version__
```

**Reply by jbailiss** [reply]
Sorry to resurrect an old conversation but I am receiving the same error message.

I have previously used Splink successfully but have now migrated to a new server which required a reinstall.  I am using Anaconda and installed using: conda install -c conda-forge splink

splink.__version__
4.0.6

duckdb.__version__
1.1.3

python_version()
3.12.3

from splink.duckdb.linker import DuckDBLinker
ModuleNotFoundError: No module named 'splink.duckdb'

I'm at a bit of a loss so any advice would be really appreciated!

Thanks


The logs from the install are below:

The following NEW packages will be INSTALLED:

  libexpat           conda-forge/win-64::libexpat-2.6.3-he0c23c2_0
  libsqlite          conda-forge/win-64::libsqlite-3.47.0-h2466b09_1
  libzlib            conda-forge/win-64::libzlib-1.2.13-h2466b09_6
  python-duckdb      conda-forge/win-64::python-duckdb-1.1.3-py312h275cf98_0
  python_abi         conda-forge/win-64::python_abi-3.12-5_cp312
  splink             conda-forge/noarch::splink-4.0.6-pyhd8ed1ab_0
  sqlglot            conda-forge/noarch::sqlglot-25.33.0-pyhd8ed1ab_0
  ucrt               conda-forge/win-64::ucrt-10.0.22621.0-h57928b3_1
  vc14_runtime       conda-forge/win-64::vc14_runtime-14.42.34433-he29a5d6_23

The following packages will be UPDATED:

  openssl              pkgs/main::openssl-3.0.15-h827c3e9_0 --> conda-forge::openssl-3.4.0-h2466b09_0
  vs2015_runtime     pkgs/main::vs2015_runtime-14.40.33807~ --> conda-forge::vs2015_runtime-14.42.34433-hdffcdeb_23
  zlib                    pkgs/main::zlib-1.2.13-h8cc25b3_1 --> conda-forge::zlib-1.2.13-h2466b09_6

The following packages will be SUPERSEDED by a higher-priority channel:

  certifi            pkgs/main/win-64::certifi-2024.8.30-p~ --> conda-forge/noarch::certifi-2024.8.30-pyhd8ed1ab_0
  conda              pkgs/main::conda-24.11.0-py312haa9553~ --> conda-forge::conda-24.11.0-py312h2e8e312_0
  python                pkgs/main::python-3.12.7-h14ffc60_0 --> conda-forge::python-3.12.3-h2628c8c_0_cpython

**Reply by RobinL** [reply]
I think your code was written for Splink 3, so you need to install to https://pypi.org/project/splink/3.9.15/ or upgrade your code to splink 4 syntax https://moj-analytical-services.github.io/splink/blog/2024/07/24/splink-400-released.html

**Reply by jbailiss** [reply]
Thanks Robin.  That was indeed the cause.

The conversion guide is very helpful and all is working on the new server now.

**Comment by py67** [comment]
I am facing a similar issue - can't import splink. I am using conda environment and list conda shows splink installed in my conda environment. Below is the message I get in importing splink. 

>>> import splink
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\yadavpk\Anaconda3\envs\nsarV\Lib\site-packages\splink\__init__.py", line 4, in <module>
    from pyspark.sql.dataframe import DataFrame
  File "C:\Users\yadavpk\Anaconda3\envs\nsarV\Lib\site-packages\pyspark\__init__.py", line 51, in <module>
    from pyspark.context import SparkContext
  File "C:\Users\yadavpk\Anaconda3\envs\nsarV\Lib\site-packages\pyspark\context.py", line 31, in <module>
    from pyspark import accumulators
  File "C:\Users\yadavpk\Anaconda3\envs\nsarV\Lib\site-packages\pyspark\accumulators.py", line 97, in <module>
    from pyspark.serializers import read_int, PickleSerializer
  File "C:\Users\yadavpk\Anaconda3\envs\nsarV\Lib\site-packages\pyspark\serializers.py", line 71, in <module>
    from pyspark import cloudpickle
  File "C:\Users\yadavpk\Anaconda3\envs\nsarV\Lib\site-packages\pyspark\cloudpickle.py", line 145, in <module>
    _cell_set_template_code = _make_cell_set_template_code()
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\yadavpk\Anaconda3\envs\nsarV\Lib\site-packages\pyspark\cloudpickle.py", line 126, in _make_cell_set_template_code
    return types.CodeType(
           ^^^^^^^^^^^^^^^
TypeError: code expected at least 16 arguments, got 15


I can see splink installed in conda environment nsarV (below is what I get):

(nsarV) PS C:\Users\yadavpk> conda list splink
# packages in environment at C:\Users\yadavpk\Anaconda3\envs\nsarV:
#
# Name                    Version                   Build  Channel
splink                    1.0.6              pyhd8ed1ab_0    conda-forge

Thanks!

**Reply by RobinL** [reply]
Hmm, not sure.  Splink 1.0.6 is a very old version of Splink.  I asked chat GPT 4 and it said:

>>> The error message "TypeError: code expected at least 16 arguments, got 15" suggests that there's a version mismatch somewhere. It seems like you are running a version of PySpark that's incompatible with the Python version you have installed in your conda environment.

>>> PySpark is a Python library for Apache Spark, a big data processing engine. It seems like the version of PySpark being used by Splink is not compatible with your Python version.


Is that a possibility?

**Reply by py67** [reply]
Your reply made me notice that the version of Splink I recently installed using Conda is indeed very old. Just installed version 3.8.1 and the problem is gone. Thanks so much!

**Comment by Nyraa1** [comment]
Hi , I am trying to do vector based algorithm in splink for deduplication but cosine similarity at threshold is not working so I have pre computed it outside splink and checking in splink. Due to this rows have increased outside splink when making it into pairs and again increasing a lot when used in splink. It is running for long time as well getting out of memory as well.

Any way how to reduce this rows or any other approach for vector based

note: using splink 4.0.7

**Reply by RobinL** [reply]
We'd need to see a reprex to be able to help i.e a runnable example that produces the errors you're getting

**Reply by Nyraa1** [reply]
Code as follows:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
import pandas as pd

# Sample data
df = pd.DataFrame({
    "unique_id": [1, 2, 3],
    "full_name": ["John Smith", "Jon Smyth", "Jane Doe"],
    "email": ["john@gmail.com", "jon@gmail.com", "jane@gmail.com"]
})

# TF-IDF for name
vectorizer = TfidfVectorizer().fit(df["full_name"])
tfidf_matrix = vectorizer.transform(df["full_name"])
cosine_sim_matrix = cosine_similarity(tfidf_matrix)

# Create pairwise scores
pairs = []
for i in range(len(df)):
    for j in range(i+1, len(df)):
        pairs.append({
            "id1": df.loc[i, "unique_id"],
            "id2": df.loc[j, "unique_id"],
            "cosine_name_score": cosine_sim_matrix[i, j],
            "full_name_l": df.loc[i, "full_name"],
            "full_name_r": df.loc[j, "full_name"]
        })
df_sim = pd.DataFrame(pairs)

settings = {
    "link_type": "dedupe_only",
    "comparisons": [
        # Cosine similarity comparisons
        {
            "output_column_name": "cosine_name_comparison",
            "comparison_levels": [
                {
                    "sql_condition": "cosine_name > 0.8",
                    "label": "High similarity"
                },
                {
                    "sql_condition": "cosine_name > 0.6",
                    "label": "Medium similarity"
                },
                {
                    "sql_condition": "cosine_name <= 0.6",
                    "label": "Low or no match"
                }
            ]
        }
    ],
    "unique_id_column_name": "id1"
}

linker code below
